keylime/config.py

```diff
@@ -73,45 +74,257 @@ def environ_bool(env_name, default):
+if "KEYLIME_VERIFIER_CONFIG" in os.environ:
+    CONFIG_ENV["verifier"] = os.environ["KEYLIME_VERIFIER_CONFIG"]
+if "KEYLIME_TENANT_CONFIG" in os.environ:
+    CONFIG_ENV["tenant"] = os.environ["KEYLIME_TENANT_CONFIG"]
+if "KEYLIME_REGISTRAR_CONFIG" in os.environ:
+    CONFIG_ENV["registrar"] = os.environ["KEYLIME_REGISTRAR_CONFIG"]
+if "KEYLIME_CA_CONFIG" in os.environ:
+    CONFIG_ENV["ca"] = os.environ["KEYLIME_CA_CONFIG"]
+if "KEYLIME_LOGGING_CONFIG" in os.environ:
+    CONFIG_ENV["logging"] = os.environ["KEYLIME_LOGGING_CONFIG"]
+
+# Single instance
+_config = None
+
+
+def get_config(component) -> configparser.RawConfigParser:
+    """Find the configuration file to use for the given component and apply the
+    overrides defined by configuration snippets.
+
+    Configuration files are expected to be installed by the distribution on
+    /usr/etc/keylime or /etc/keylime. If a configuration file is found in
+    /etc/keylime, the configuration file in /usr/etc/keylime is ignored.
+
+    If a configuration file path is set through a KEYLIME_*_CONFIG environment
+    variable, all configuration from other files for that component are ignored,
+    meaning that the configuration file set through a environment variable has
+    top priority.
+
+    The system administrator can define overrides for the values through
+    configuration snippets in /etc/keylime/<component>.config.d, where
+    <component> is one of: "agent", "verifier", "tenant", "registrar", "ca", or
+    "logging".
+
+    The configuration processing follows the steps described below:
+
+    * If a configuration path is set through environment variable, use the
+    configuration from this file and ignore configuration from other files.
+    * Check if the configuration file for the component is present in
+    /usr/etc/keylime (e.g. /usr/etc/keylime/agent.conf). If found, this file is
+    used as the base configuration file for the component
+    * Check if the configuration file for the component is present in
+    /etc/keylime (e.g. /etc/keylime/agent.conf). If found, set this file as the
+    base configuration file for the component, ignoring any previously set base
+    configuration file.
+    * Find and apply any override from files in /etc/keylime/<component.d> to
+    the base configuration
+    * Find and apply any local user override from files in
+    ~/config/keylime/<component.d>
+    """
+
+    global _config
+
+    if not _config:
+        _config = {}
+
+    if not component:
+        # If no component was specified, return an empty configuration
+        raise Exception("No component provided to get_config")
+
+    if component not in _config:  # pylint: disable=too-many-nested-blocks
         # Use RawConfigParser, so we can also use it as the logging config
-        config = configparser.RawConfigParser()
+        _config[component] = configparser.RawConfigParser()
+
+        if not CONFIG_ENV or not isinstance(CONFIG_ENV, dict):
```

Comment:
Why here, and more later, are we validating a variable that is expected to be "static const" and declared only for this file?

Defensive programming is OK but if we do not expect that this can be changed, maybe is too much?

Reply:
To be honest, I only added these checks because `test_config.py` had test cases for such changes.